polkadot/runtime/parachains/src/util.rs

```diff
@@ -115,6 +118,68 @@ pub fn take_active_subset<T: Clone>(active: &[ValidatorIndex], set: &[T]) -> Vec
 	subset
 }
 
+pub(crate) trait PopulateKeys {
+	type Key;
+	type Value;
+
+	/// Load from actual storage into a HashMap. We need to ensure there are no
+	/// hash collisions
+	fn populate() -> HashMap<Self::Key, Self::Value>;
+}
+
+// An in memory overlay for large storage maps that we frequently iterate and want
+// to reading/writing to multiple times. A cache with write back capabilities.
+//
+// Maybe TODO(in the future): move this capability to FRAME and ensure we use blake256 hashing of
+// keys, to ensure attackers cannot control the keys and force us to create overweight blocks.
+pub(crate) struct StorageMapOverlay<K, V, P: PopulateKeys> {
+	data: hashbrown::HashMap<K, V>,
+	modified: hashbrown::HashSet<K>,
+	_phantom: PhantomData<P>,
+}
+
+impl<K, V, P> StorageMapOverlay<K, V, P>
+where
+	K: sp_std::hash::Hash + Eq + PartialEq + Clone + Copy,
+	V: Clone,
+	P: PopulateKeys<Key = K, Value = V>,
+{
+	// Construct a new overlay instance.
+	pub fn new() -> Self {
+		let data = P::populate();
```

Comment:
Given that we are doubling the memory footprint here, it would be really good if this improved performance by more than just a few percent.

Reply:
Right now it improves it by around 12% . I'd say it is not great, but that's what is currently possible. What kind of improvement are you expecting for the tradeoff to be worth it ?

Reply:
I'll think about it a bit. In the first mutate, we update the availability votes. But when doing the enactments, we may enact more candidates than the ones that have been just made available in this block.
If for example we have A B C backed. B and C are already available. A is just being made available now. So we have to enact all candidates (A B and C).

That's why I chose to iterate twice there. I'll think more about it to see if we can call mutate only once for each para